### ISSUE TITLE
[ci] workflow to build package with GPU support

### ISSUE THREAD

**Issue by hero-intelligent** [karaokenerds/python-audio-separator#10]
>  if anyone has a way to make this cleaner so we can support both CPU and CUDA transcodes without separate installation processes, please let me know or submit a PR!

I've made a Dockerfile of UVR5 th if anyone has a way to make this cleaner so we can support both CPU and CUDA transcodes without separate installation processes, please let me know or submit a PR!e day before yesterday, and i've posted the Dockerfile in an issue, in case anyone should need it sometime. https://github.com/Anjok07/ultimatevocalremovergui/issues/379#issuecomment-1694905668

This idea was raised when I took my first step into docker, or say programming as a hobby, in order not messing up with or breaking my system, so that I needn't reinstall my system again and again, because i was doing it on my only computer that is too powerless to spin up a virtual machine. But now, that computer was finally broken, so I've changed one a little bit powerful. Anyways, that goes all into my history.

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Now, this is my thought that you should take an advantage of this image. Later, i'll push this giant 10GB docker image to dockerhub for you to test out.

My computer I'm using now is without GPU. So, in other words, if the image built with cuda on a machine that is without nvidia GPU installed can work perfectly and perfectly on a machine with an nvidia GPU installed, certain methodology can be applied to build the package.

 If anyone with GPU can do us a favor to test if he / she can execute the commands that are listed below **INSIDE THE CONTAINER** and get the correct output, you may modify your working flow using nvidia docker image to build your package.

```bash
# This should get the output
nvidia-smi

# This should spit out "True"
python -c "import torch; result = torch.cuda.is_available(); print(result)" 
```

So, if it is a success, maybe in theory, you could just change `[tool.poetry.dependencies.onnxruntime]` into `[tool.poetry.dependencies.onnxruntime-gpu]` in`pyproject.toml` and run your workflow using docker, if the CPU version and GPU version of onnxruntime is compatible. 

But it certainly needs volunteers for further testing, so this should not be pushed to pypi yet, or use an alpha tag. 

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

There is another possibility to test out. 

Could you just build your package mannually from source? 

First do it on a virtual machine instance that is with cuda installed, then instead of pushing to and then pull from pypi, download the package to your own computer, then destroy the instance.

Second, spin up a similar one and upload the freshly built package to the new instance and test if it works properly. 

If it does, yes, your program source code has no bug, the only thing that needs worry about is the environment; if not, that's your code's problem. 

Third, spin up a virtual machine on your own computer, which is certainly without GPU installed, and test if the freshly built package can work properly without GPU. This is a test to find out whether the onnxruntime-gpu can run on a machine without GPU installed.

----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

I'm looking forward to your success. 

best wishes

**Comment by beveradb** [comment]
Thanks for sharing 😄 
I certainly like the idea of having a docker image ready to run for anyone who wants to use audio-separator without any installation required; that should also make it eas

However, I gotta say that dockerfile you linked is pretty wild 😅 
Not sure I'd want to use it directly but there's probably some learnings I could take from it and try to publish a docker image from this repo for people to be able to conveniently run `audio-separator` in docker.

My gut feeling is that should be possible with a much more minimal dockerfile/image - e.g. why are you compiling python from source?
Why not just start with a base image using the [official pytorch images](https://hub.docker.com/r/pytorch/pytorch/tags) (e.g. the latest CUDA one [2.0.1-cuda11.7-cudnn8-runtime](https://hub.docker.com/layers/pytorch/pytorch/2.0.1-cuda11.7-cudnn8-runtime/images/sha256-82e0d379a5dedd6303c89eda57bcc434c40be11f249ddfadfd5673b84351e806?context=explore) )?

**Comment by beveradb** [comment]
Hey @hero-intelligent, sorry for the slow reply on this, didn't have much capacity for hobby projects till now but i'm focusing on `audio-separator` a bit more over the festive period.

Just a heads up, I've just pushed a bunch of updates to `audio-separator` to make the GPU/other hardware acceleration detection a lot more robust and remove the need for passing in the `--use_cuda` or `--use_coreml` arguments.

I've also repackaged it to relax the dependencies and remove the CPU version of onnxruntime from the dependencies; in fact no version of onnxruntime is a default dependency any more, instead people have to install `audio-separator` with a specific extra (either `gpu`, `silicon`, or `cpu`)

So for GPU users, in theory all most people should need to do is install `audio-separator` like this:
`pip install audio-separator[gpu]`

Check out the updated readme:
https://github.com/karaokenerds/python-audio-separator#-nvidia-gpu-with-cuda-acceleration

I've tested this pretty robustly on a Runpod with a RTX A4000 GPU but I may still have missed something. 
Please let me know if you have any issues!

**Comment by beveradb** [comment]
I'm going to close this issue for now as I believe the `audio-separator[gpu]` package on PyPI works as expected out of the box for a CUDA user - but please feel free to re-open and share your logs (ideally with `--log_level debug`) here for me to investigate if it doesn't work for you!
